mountpoint-s3/src/data_cache/cache_directory.rs

```diff
@@ -29,66 +36,102 @@ pub enum ManagedCacheDirError {
-    /// If the directory already exists, it will be deleted before being recreated.
-    pub fn new_from_parent<P: AsRef<Path>>(parent_path: P) -> Result<Self, ManagedCacheDirError> {
+    /// If `<parent_path>/mountpoint-cache` already exists, it will be deleted before being
+    /// recreated. This can cause performance degradation, but will never result in unused caches
+    /// being retained, assuming other Mountpoint instances are being ran on the host.
+    pub fn new_from_parent_with_cache_key(
+        parent_path: impl AsRef<Path>,
+        cache_key: Option<OsString>,
+    ) -> Result<Self, ManagedCacheDirError> {
+        let mountpoint_cache_path = parent_path.as_ref().join("mountpoint-cache");
+        let managed_cache_path = match cache_key {
+            None => mountpoint_cache_path.clone(),
+            Some(cache_key) => mountpoint_cache_path.join(hash_cache_key(cache_key.as_bytes())),
+        };
         let managed_cache_dir = Self {
-            managed_path: parent_path.as_ref().join("mountpoint-cache"),
+            mountpoint_cache_path,
+            managed_cache_path,
         };
 
         managed_cache_dir.remove()?;
-
-        let mkdir_result = fs::DirBuilder::new().mode(0o700).create(managed_cache_dir.as_path());
-        if let Err(mkdir_err) = mkdir_result {
-            match mkdir_err.kind() {
-                io::ErrorKind::AlreadyExists => tracing::warn!(
-                    cache_dir = ?managed_cache_dir.as_path(),
-                    "cache sub-directory already existed immediately after removal",
-                ),
-                _kind => return Err(ManagedCacheDirError::CreationFailure(mkdir_err)),
-            }
-        }
-
+        managed_cache_dir.create_cache_dir()?;
         Ok(managed_cache_dir)
     }
 
     /// Remove the cache sub-directory, along with its contents if any
     fn remove(&self) -> Result<(), ManagedCacheDirError> {
-        tracing::debug!(cache_subdirectory = ?self.managed_path, "removing the cache sub-directory and any contents");
-        if let Err(remove_dir_err) = fs::remove_dir_all(&self.managed_path) {
+        tracing::debug!(cache_subdirectory = ?self.mountpoint_cache_path, "removing the cache sub-directory and any contents");
+        if let Err(remove_dir_err) = fs::remove_dir_all(&self.mountpoint_cache_path) {
             match remove_dir_err.kind() {
                 io::ErrorKind::NotFound => (),
                 _kind => return Err(ManagedCacheDirError::CleanupFailure(remove_dir_err)),
             }
         }
-        tracing::trace!(cache_subdirectory = ?self.managed_path, "cache sub-directory removal complete");
+        tracing::trace!(cache_subdirectory = ?self.mountpoint_cache_path, "cache sub-directory removal complete");
+        Ok(())
+    }
+
+    /// Create the cache sub-directory, assumes the parent path exists.
+    fn create_cache_dir(&self) -> Result<(), ManagedCacheDirError> {
```

Comment:
nit: I'd rather inline this code in `new_from_parent_with_cache_key`, so all the logic to deal with the 2 levels is in a single place.